src/Temporalio/Activity/ActivityContext.cs

```diff
@@ -0,0 +1,122 @@
+
+        /// <summary>
+        /// Gets the info for this activity.
+        /// </summary>
+        public ActivityInfo Info { get; private init; }
+
+        /// <summary>
+        /// Gets the logger scoped to this activity.
+        /// </summary>
+        public ILogger Logger { get; private init; }
+
+        /// <summary>
+        /// Gets why the activity was cancelled. This value is inaccurate until
+        /// <see cref="CancellationToken" /> is cancelled.
+        /// </summary>
+        public ActivityCancelReason CancelReason => CancelReasonRef.CancelReason;
+
+        /// <summary>
+        /// Gets the cancellation token that is cancelled when the activity is cancelled.
+        /// </summary>
+        public CancellationToken CancellationToken { get; private init; }
+
+        /// <summary>
+        /// Gets the cancellation token that is cancelled when the worker is shutdown. This can be
+        /// used to gracefully handle worker shutdown for
+        /// <see cref="Worker.TemporalWorkerOptions.GracefulShutdownTimeout" /> before
+        /// <see cref="CancellationToken" /> will ultimately be cancelled.
+        /// </summary>
+        public CancellationToken WorkerShutdownToken { get; private init; }
+
+        /// <summary>
+        /// Gets the async local current value.
+        /// </summary>
+        internal static AsyncLocal<ActivityContext?> AsyncLocalCurrent { get; } = new();
+
+        /// <summary>
+        /// Gets or sets the heartbeater. This is late-bound since interceptors that need this
+        /// context may be invoked before the interceptor is created that ends up providing the
+        /// heartbeater.
+        /// </summary>
+        internal Action<object?[]>? Heartbeater { get; set; }
+
+        /// <summary>
+        /// Gets a reference to the reason enum.
+        /// </summary>
+        internal ActivityCancelReasonRef CancelReasonRef { get; init; } = new();
+
+        /// <summary>
+        /// Gets the raw proto task token for this activity.
+        /// </summary>
+        internal ByteString TaskToken { get; private init; }
+
+        /// <summary>
+        /// Record a heartbeat on the activity.
+        /// </summary>
+        /// <remarks>
+        /// Heartbeats should be used for all non-immediately-returning, non-local activities and
+        /// they are required to receive cancellation. Heartbeats are queued and processed
+        /// asynchronously, so this will not error if the details cannot be converted. Rather any
+        /// error converting heartbeat details will result in activity cancellation then activity
+        /// failure.
+        /// </remarks>
+        /// <param name="details">Details to record with the heartbeat if any.</param>
+        public void Heartbeat(params object?[] details)
+        {
+            if (Heartbeater == null)
+            {
+                throw new InvalidOperationException("Heartbeater not set yet");
+            }
```

Comment:
No, from the docs on `Heartbeater`:

> Gets or sets the heartbeater. This is late-bound since interceptors that need this context may be invoked before the interceptor is created that ends up providing the heartbeater.

So basically on "intercept activity" call I make this context available without heartbeating